### PR TITLE
test: port shell install CNI test to Go

### LINKED ISSUE #34: Rework test-install-cni.sh in golang
Implement and enhance the logic in deployments/kubernetes/install/test/test-install-cni.sh in golang in test/install_k8s_test.go.  

The golang code could make use of a more intelligent config comparison with libs like jsondiff or custom validation.

### PR DESCRIPTION
**Pull request by venilnoronha**
This PR ports the existing install CNI test from shell to Golang. Fixes #34.

### CODE REVIEW

**Comment by venilnoronha** [comment]
I haven't got a chance to test this yet.

**Comment by tiswanso** [comment]
@venilnoronha -- I'll try to make time to give it a test but I was hoping circleci hookup would be ready by now.  Still seems to be some issue with that.  When that is working it'd auto-run in the circleci install-cni job.

Initial code read-through looks good.

**Comment by venilnoronha** [comment]
@tiswanso thanks! I'll rebase once CircleCI is up.

**Comment by john-a-joyce** [comment]
@venilnoronha circleci should be properly hooked up now, but you probably have to add a commit to trigger it

**Comment by john-a-joyce** [comment]
LGTM

File: deployments/kubernetes/install/test/install_cni.go [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+	containerID, err := cmd.Output()
+	if err != nil {
+		t.Fatalf("Test %v ERROR: failed to start docker container '%v', see %v",
+			testNum, dockerImage, errFileName)
+	}
+	t.Logf("Container ID: %s", containerID)
+	return strings.Trim(string(containerID), "\n")
+}
+
+func docker(cmd, containerID string, t *testing.T) {
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply deployments/kubernetes/install/test/install_cni.go]: The one of line 152 pipes error output to a file and this function writes it to `STDOUT`. That's why the difference.

We can actually create a separate PR for restructuring the test in this repository and address this while building a framework with these utilities.

File: deployments/kubernetes/install/test/install_cni.go [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		for _, f := range files {
+			if _, err := os.Stat(tempCNIBinDir + "/" + f); !os.IsNotExist(err) {
+				t.Fatalf("FAIL: File %v was not removed from %v", f, tempCNIBinDir)
+			} else {
+				t.Logf("PASS: File %v was removed from %v", f, tempCNIBinDir)
+			}
+		}
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply deployments/kubernetes/install/test/install_cni.go]: Addressed in 465508c.

File: deployments/kubernetes/install/test/install_cni_test.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+var (
+	preConfFlag        = flag.String("preconf", "", "pre_conf")
+	resultFileNameFlag = flag.String("resultfilename", "", "result_filename")
+	expectedConfFlag   = flag.String("expectedconf", "", "expected_conf")
+	expectedCleanFlag  = flag.String("expectedclean", "", "expected_clean")
+)
+
+func TestInstallCNI(t *testing.T) {
+	RunInstallCNITest(1, *preConfFlag, *resultFileNameFlag, *expectedConfFlag, *expectedCleanFlag, t)
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply deployments/kubernetes/install/test/install_cni_test.go]: Addressed in 465508c.

File: deployments/kubernetes/install/test/install_cni.go [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+	} else {
+		tempFail := mktemp(testWorkRootDir, result+".fail.XXXX", t)
+		t.Errorf("FAIL: result doesn't match expected: %v v. %v", tempResult, expected)
+		cp(tempResult, tempFail, t)
+		t.Fatalf("Check %v for diff contents", tempFail)
+	}
+}
+
+func checkBinDir(t *testing.T, tempCNIBinDir, op string, files ...string) {
+	if op == "add" {
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply deployments/kubernetes/install/test/install_cni.go]: Addressed in 465508c.

File: deployments/kubernetes/install/test/install_cni.go [hunk 5]
```diff
@@ -0,0 +1,272 @@
+// Copyright 2018 Istio authors
```
[the remark this answers is not in the document]
Comment (venilnoronha) [reply deployments/kubernetes/install/test/install_cni.go]: Addressed in 465508c.

**Comment by john-a-joyce** [comment]
LGTM

File: deployments/kubernetes/install/test/install_cni.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (venilnoronha) [reply deployments/kubernetes/install/test/install_cni.go]: Tracked via #60.

**Pull request merged**